<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>ImmuneSignatures2Data.knit</title>

<script src="ImmuneSignatures2Data_files/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="ImmuneSignatures2Data_files/bootstrap-3.3.5/css/bootstrap.min.css" rel="stylesheet" />
<script src="ImmuneSignatures2Data_files/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="ImmuneSignatures2Data_files/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="ImmuneSignatures2Data_files/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="ImmuneSignatures2Data_files/navigation-1.1/tabsets.js"></script>
<link href="ImmuneSignatures2Data_files/highlightjs-9.12.0/default.css" rel="stylesheet" />
<script src="ImmuneSignatures2Data_files/highlightjs-9.12.0/highlight.js"></script>
<link href="ImmuneSignatures2Data_files/pagedtable-1.1/css/pagedtable.css" rel="stylesheet" />
<script src="ImmuneSignatures2Data_files/pagedtable-1.1/js/pagedtable.js"></script>

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>








<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
pre code {
  padding: 0;
}
</style>



<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->




</head>

<body>


<div class="container-fluid main-container">




<div id="header">




</div>


<div id="the-immune-signatures-data-resource-a-compendium-of-systems-vaccinology-datasets" class="section level1">
<h1>The Immune Signatures Data Resource: A compendium of systems vaccinology datasets</h1>
<p>Vaccines are among the most cost-effective public health interventions for preventing infection-induced morbidity and mortality, yet much remains to be learned regarding mechanisms. The NIH/NIAID Human Immunology Project Consortium(HIPC) has leveraged systems immunology approaches to identify molecular signatures of immunogenicity in many vaccines. These data are available to the scientific community through the ImmuneSpace data portal and analysis engine leveraging the NIH/NIAID ImmPort repository. However, a barrier to progress in this area is that comparative vaccine analyses have been limited by the distributed nature of some data, potential confounders, and the absence of relevant studies in ImmPort. To support comparative analysis, we have created the Immune Signatures Data Resource, a compendium of standardized systems vaccinology datasets available through ImmuneSpace, along with code to reproduce the standardization pipeline. The current release comprises 1405 participants from 53 cohorts profiling the response to 24 vaccines and includes transcriptional profiles and antibody response measurements. This novel compendium represents a valuable resource for comparative analyses that will accelerate our understanding of mechanisms underlying vaccine responses.</p>
<p><a href="https://www.biorxiv.org/content/10.1101/2021.11.05.465336v1">Read the preprint</a></p>
<div id="all-data" class="section level2">
<h2>All Data</h2>
<p><a href="/_webdav/HIPC/IS2/@files/immunesignatures2.zip"><button>Download immunesignatures.zip (3.96 GB)</button></a></p>
<p><strong>DOI:</strong> <a href="https://doi.org/10.6084/m9.figshare.17096978.v1">10.6084/m9.figshare.17096978</a></p>
<p><strong>Note:</strong> The ImmuneSignatures data resource is derived from publicly available data on ImmPort, and available for reuse under the ImmPort <a href="https://www.immport.org/agreement">data use agreement</a>.</p>
<p><strong>Data Format</strong>: Most of the datasets are presented here as <code>.rds</code> files, that is, R objects which can be read into an R session via <code>readRDS()</code></p>
</div>
<div id="expression-sets" class="section level2">
<h2>Expression Sets</h2>
<div class="is2-downloads table-condensed">

</div>
<div id="other-datasets" class="section level3">
<h3>Other Datasets</h3>
<div class="is2-downloads table-condensed">

</div>
<div id="preview" class="section level4">
<h4>Preview</h4>
</div>
</div>
</div>
<div id="raw-data" class="section level2">
<h2>Raw data</h2>
<p><a href="/study/HIPC/IS2/dataset.view?datasetId=5014">Gene expression data files</a></p>
<div id="immune-response-data" class="section level3">
<h3>Immune Response Data</h3>
<p><a href="/study/HIPC/IS2/dataset.view?datasetId=5004">Enzyme-Linked ImmunoSpot (ELISPOT)</a><br> <a href="/study/HIPC/IS2/dataset.view?datasetId=5005">Hemagglutination inhibition (HAI)</a><br> <a href="/study/HIPC/IS2/dataset.view?datasetId=5009">Neutralizing antibody titer</a><br></p>
<p>Immune Response Data for all assays is also available as an R object:</p>
<p><a href="/_webdav/HIPC/IS2/@files/data/immdata_all.rds">immdata_all.rds</a><br></p>
<hr />
</div>
</div>
<div id="code" class="section level2">
<h2>Code</h2>
<div id="manuscript-figures" class="section level3">
<h3>Manuscript Figures</h3>
<p><a href="/_webdav/HIPC/IS2/%40files/html_outputs/prevax.html" target="_blank"><strong>Reproduction of manuscript figures</strong></a></p>
</div>
<div id="preprocessing" class="section level3">
<h3>Preprocessing</h3>
<p>Rmarkdown outputs from preprocessing, including code and some summaries. Preprocessing depends on the <a href="https://github.com/RGLab/ImmuneSignatures2">ImmuneSignatures2</a> R package. (DOI: <a href="https://zenodo.org/record/5706261#.YaVs1caIZQI">10.5281/zenodo.5706261</a>)</p>
<ol style="list-style-type: decimal">
<li><a href="/_webdav/HIPC/IS2/%40files/html_outputs/pull_esets.html" target="_blank">Pull Expressionsets</a>
<ul>
<li>Pulls gene expression data from ImmuneSpace server, constructs expressionsets, and saves output to an R object on disk.</li>
</ul></li>
<li><a href="/_webdav/HIPC/IS2/%40files/html_outputs/generate_base_eset.html" target="_blank">Generate Base ExpressionSet</a>
<ul>
<li>Perform basic QC and filtering, and save output to one ExpressionSet.</li>
</ul></li>
<li><a href="/_webdav/HIPC/IS2/%40files/html_outputs/create_final_esets.html" target="_blank">Create Final ExpressionSets</a>
<ul>
<li>Perform cross-study normalization and batch correction, and save the different versions available in the table above.</li>
</ul></li>
</ol>
<style>


  .is2-downloads td, .is2-downloads th {
    border: 1px solid #ddd;
    padding: 3px;
  }
  .is2-downloads tr:nth-child(even){background-color: #f2f2f2;}
  .is2-downloads tr:hover {background-color: #ddd;}
  .is2-downloads th {
    background-color: #255c97;
    text-align: left;
    color: white;
    font-size: 15px;
    font-weight: bold;
    padding: 3px;
  }

</style>
</div>
</div>
</div>




</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
