# Differential expression analysis after polyIC treatment in the VL+/- cohorts

### Report generated by `r labkey.user.email`
```{r knitr-options, echo=FALSE, cache=FALSE}
library(knitr)
opts_chunk$set(echo=TRUE, eval=TRUE, message=FALSE, warning=FALSE, error=TRUE)
opts_chunk$set(cache=FALSE, cache.path=paste(labkey.file.root, "cache/limma_analysis_polyIC/", sep="/"))
options(width=80)
```

```{r load-required-packages}
library(Biobase)
library(limma)
library(Rlabkey)
library(data.table)
library(pheatmap)
```

```{r expression, cache=TRUE}
#read gene_expression_files
exMat <- labkey.selectRows(baseUrl = labkey.url.base, folderPath = labkey.url.path,
      schemaName = "study", queryName = "gene_expression_matrices", colNameOpt = "rname")
exMatLink <- paste(labkey.file.root, "analysis/exprs_matrices", unique(exMat$file_info_name), sep="/")
exprs <- read.table(exMatLink)
```

### Get the probe mapping file
```{r fdata, cache=TRUE, dependson="expression"}
fdLink <- paste(labkey.file.root, "analysis/features2genes", unique(exMat$feature_mapping_file), sep="/")
fdata <- read.table(fdLink, header=TRUE, sep="\t")
```

### Get the phenotypic data
```{r expsamples}
bs_filter <- makeFilter(c("BIOSAMPLE_ACCESSION", "IN", paste(unique(exMat$biosample_accession), collapse=";")))
bs_2_es <- labkey.selectRows(baseUrl = labkey.url.base, folderPath = labkey.url.path,
                             schemaName = "immport", queryName = "biosample_2_expsample",
                             colFilter = bs_filter, colNameOpt = "rname")
```

```{r treatments}
es_filter <- makeFilter(c("EXPSAMPLE_ACCESSION", "IN", paste(bs_2_es$expsample_accession, collapse=";")))
es_2_trt <- labkey.selectRows(baseUrl = labkey.url.base, folderPath = labkey.url.path,
                              schemaName = "immport", queryName = "expsample_2_treatment",
                              colFilter = es_filter, colNameOpt = "rname")
trt_filter <- makeFilter(c("TREATMENT_ACCESSION", "IN", paste(es_2_trt$treatment_accession, collapse=";")))
trt <- labkey.selectRows(labkey.url.base, labkey.url.path, "immport", "treatment",
                         colFilter = trt_filter, colNameOpt = "rname")
```

```{r cell-type}
bs_2_ptl <- labkey.selectRows(labkey.url.base, labkey.url.path, "immport", "biosample_2_protocol",
                              colFilter = bs_filter, colNameOpt = "rname")

ptl_filter <- makeFilter(c("PROTOCOL_ACCESSION", "IN", paste(bs_2_ptl$protocol_accession, collapse=";")))
ptl <- labkey.selectRows(labkey.url.base, labkey.url.path, "immport", "protocol",
                              colFilter = ptl_filter, colNameOpt = "rname")
```

```{r merge-exmat}
pd <- bs_2_es[, c("biosample_accession", "expsample_accession")]
pd$treatment_accession <- es_2_trt[match(bs_2_es$expsample_accession, es_2_trt$expsample_accession),"treatment_accession"]
pd$treatment_name <- trt[match(pd$treatment_accession, trt$treatment_accession), "name"]
pd$protocol_accession <- bs_2_ptl[match(pd$biosample_accession, bs_2_ptl$biosample_accession), "protocol_accession"]
pd$protocol_name <- ptl[match(pd$protocol_accession, ptl$protocol_accession),"name"]

exMat$treatment <- pd[match(exMat$biosample_accession, pd$biosample_accession), "treatment_name"]
exMat$protocol <- pd[match(exMat$biosample_accession, pd$biosample_accession), "protocol_name"]
rownames(exMat) <- exMat$biosample_accession
exMat <- exMat[match(colnames(exprs), exMat$biosample_accession),]
```


## Create an expressionSet
```{r eset, cache=TRUE, dependson="expression;fdata"}
eset <- ExpressionSet(assayData = as.matrix(exprs), phenoData = as(exMat, "AnnotatedDataFrame"), featureData = as(fdata, "AnnotatedDataFrame"))
```

```{r subset-data, cache=TRUE, dependson="eset"}
eset <- eset[,grep("macro", tolower(eset$protocol))]
pData(eset) <- droplevels(pData(eset))
eset$subject_accession <- as.factor(eset$subject_accession)
eset$treatment <- as.factor(eset$treatment)
```

## Differential expression analysis using limma
```{r filter-opts}
FDRthresh <- 0.05
FCthresh <- 1.5
```

### polyIC responsive genes:
```{r limma-polyIC, cache=TRUE, dependson='eset;filter-opts;subset-data'}
mm <- model.matrix(~subject_accession + treatment, eset)
fit <- lmFit(eset, mm)
ebay <- eBayes(fit)
tt_polyIC <- topTable(ebay, coef = "treatmentTR_HCV_PolyICH", number = Inf)
polyIC_probes <- tt_polyIC[tt_polyIC$adj.P.Val < FDRthresh & abs(tt_polyIC$logFC) > log2(FCthresh), ]
up <- polyIC_probes[polyIC_probes$logFC>0,]
down <- polyIC_probes[polyIC_probes$logFC<0,]
```

Using limma with an FDR cutoff of `r FDRthresh` and an absolute fold-change threshold of `r FCthresh`, we find `r nrow(up)` up- and `r nrow(down)` down-regulated probes in response to the polyIC treatment.  

### Differential expression analysis of the two cohorts:
```{r limma-cohort-effect, cache=TRUE, dependson='eset;filter-opts'}
eset <- eset[, order(eset$subject_accession, eset$treatment)]
FC <- exprs(eset)[,seq(2,length(sampleNames(eset)),2)] - exprs(eset)[,seq(1,length(sampleNames(eset)),2)]
sseset <- eset[,seq(2,length(sampleNames(eset)),2)]
mm <- model.matrix(~arm_name, sseset)
fit <- lmFit(sseset, mm)
ebay <- eBayes(fit)
tt_hcv <- topTable(ebay, coef = 2, number = Inf)
hcvPos_probes <- tt_hcv[tt_hcv$adj.P.Val < FDRthresh & abs(tt_hcv$logFC) > log2(FCthresh), ]
up <- hcvPos_probes[hcvPos_probes$logFC>0,]
down <- hcvPos_probes[hcvPos_probes$logFC<0,]
```
Using limma with an FDR cutoff of `r FDRthresh` and an absolute fold-change threshold of `r FCthresh`, we find `r nrow(up)` up- and `r nrow(down)` down-regulated probes in patients with chronic genotype 1 HCV infection. 

<br>
We can now write these files to disk as follows:
```{r save-results, cache=TRUE, dependson="limma-polyIC;limma-cohort-effect", eval=TRUE}
limma_table <- tt_polyIC[,c("feature_id", "gene_symbol", "logFC", "AveExpr", "t", "P.Value", "adj.P.Val")]
setnames(limma_table,c("logFC", "AveExpr", "t", "P.Value", "adj.P.Val"), c("log_fc", "ave_expr", "statistic", "p_value", "adj_p_val"))
limma_table$analysis_accession <- "GEA8"
write.table(limma_table, paste0(labkey.file.root, "/analysis/gene_expression_analysis/limma/limma_table_GEA8.tsv"), 
    sep = "\t", quote = FALSE, row.names = FALSE)

limma_table <- tt_hcv[,c("feature_id", "gene_symbol", "logFC", "AveExpr", "t", "P.Value", "adj.P.Val")]
setnames(limma_table,c("logFC", "AveExpr", "t", "P.Value", "adj.P.Val"), c("log_fc", "ave_expr", "statistic", "p_value", "adj_p_val"))
limma_table$analysis_accession <- "GEA9"
write.table(limma_table, paste0(labkey.file.root, "/analysis/gene_expression_analysis/limma/limma_table_GEA9.tsv"), 
    sep = "\t", quote = FALSE, row.names = FALSE)
```
